pkgs/by-name/op/opencl-clang/llvm/clang/default.nix

```diff
@@ -0,0 +1,161 @@
+  cmake,
+  ninja,
+  libxml2,
+  libllvm,
+  version,
+  python3,
+  tblgen,
+  replaceVars,
+  fetchpatch,
+}:
+stdenv.mkDerivation (finalAttrs: {
+  pname = "clang";
+  inherit version;
+
+  src = runCommand "clang-src-${version}" { inherit (monorepoSrc) passthru; } ''
+    mkdir -p "$out"
+    cp -r ${monorepoSrc}/cmake "$out"
+    cp -r ${monorepoSrc}/clang "$out"
+    cp -r ${monorepoSrc}/clang-tools-extra "$out"
+  '';
+
+  sourceRoot = "${finalAttrs.src.name}/clang";
+
+  patches = [
+    ./purity.patch
+    # Remove extraneous ".a" suffix from baremetal clang_rt.builtins when compiling for baremetal.
+    # https://reviews.llvm.org/D51899
+    ./gnu-install-dirs.patch
+
+    # https://github.com/llvm/llvm-project/pull/116476
+    # prevent clang ignoring warnings / errors for unsuppored
+    # options when building & linking a source file with trailing
+    # libraries. eg: `clang -munsupported hello.c -lc`
+    ./clang-unsupported-option.patch
+    (fetchpatch {
+      name = "ignore-nostd-link.patch";
+      url = "https://github.com/llvm/llvm-project/commit/5b77e752dcd073846b89559d6c0e1a7699e58615.patch";
+      relative = "clang";
+      hash = "sha256-qzSAmoGY+7POkDhcGgQRPaNQ3+7PIcIc9cZuiE/eLkc=";
+    })
+    # Pass the correct path to libllvm
+    (replaceVars (./clang-11-15-LLVMgold-path.patch) {
+      libllvmLibdir = "${libllvm.lib}/lib";
+    })
+
+    # Backport version logic from Clang 16. This is needed by the following patch.
+    (fetchpatch {
+      name = "clang-darwin-Use-consistent-version-define-stringifying-logic.patch";
+      url = "https://github.com/llvm/llvm-project/commit/60a33ded751c86fff9ac1c4bdd2b341fbe4b0649.patch?full_index=1";
+      includes = [ "lib/Basic/Targets/OSTargets.cpp" ];
+      stripLen = 1;
+      hash = "sha256-YVTSg5eZLz3po2AUczPNXCK26JA3CuTh6Iqp7hAAKIs=";
+    })
+    # Backport `__ENVIRONMENT_OS_VERSION_MIN_REQUIRED__` support from Clang 17.
+    # This is needed by newer SDKs (14+).
+    (fetchpatch {
+      name = "clang-darwin-An-OS-version-preprocessor-define.patch";
+      url = "https://github.com/llvm/llvm-project/commit/c8e2dd8c6f490b68e41fe663b44535a8a21dfeab.patch?full_index=1";
+      includes = [ "lib/Basic/Targets/OSTargets.cpp" ];
+      stripLen = 1;
+      hash = "sha256-Vs32kql7N6qtLqc12FtZHURcbenA7+N3E/nRRX3jdig=";
+    })
+  ];
+
+  nativeBuildInputs = [
+    cmake
+    python3
+    ninja
+  ];
+
+  buildInputs = [
+    libxml2
+    libllvm
+  ];
+
+  cmakeFlags = [
+    (lib.cmakeFeature "CLANG_INSTALL_PACKAGE_DIR" "${placeholder "dev"}/lib/cmake/clang")
+    (lib.cmakeBool "CLANGD_BUILD_XPC" false)
+    (lib.cmakeBool "LLVM_ENABLE_RTTI" true)
+    (lib.cmakeFeature "LLVM_TABLEGEN_EXE" "${tblgen}/bin/llvm-tblgen")
+    (lib.cmakeFeature "CLANG_TABLEGEN" "${tblgen}/bin/clang-tblgen")
+    (lib.cmakeFeature "CLANG_TIDY_CONFUSABLE_CHARS_GEN" "${tblgen}/bin/clang-tidy-confusable-chars-gen")
+    (lib.cmakeFeature "CLANG_PSEUDO_GEN" "${tblgen}/bin/clang-pseudo-gen")
```

Comment:
These are there to speed up cross in the presence of a cached build platform LLVM. There’s no use for the `tblgen.nix` hack for this package; it only adds complexity.